tree/treeplayer/inc/TTreeReaderArray.h

```diff
@@ -59,90 +59,159 @@ namespace Internal {
+      Iterator_t operator++(int)
+      {
+         auto ret = *this;
+         auto &self = *this;
+         ++self;
          return ret;
       }
 
-      Iterator_t& operator++() {
-         // Pre-increment (++it).
-         if (IsValid()) {
-            ++fIndex;
-            if (fIndex >= fArray->GetSize()) {
-               // Remember that it's past-end.
-               fArray = 0;
-            }
-         }
+      /// Pre-decrement operator
+      Iterator_t &operator--()
+      {
+         if (fIndex == 0u)
+            fArray = nullptr; // invalidate iterator
+         else
+            --fIndex;
          return *this;
       }
 
-      T& operator*() const {
-         // Get the referenced element.
-         R__ASSERT(fArray && "invalid iterator!");
-         return fArray->At(fIndex);
+      /// Post-decrement operator
+      Iterator_t operator--(int)
+      {
+         auto ret = *this;
+         auto &self = *this;
+         --self;
+         return ret;
       }
 
-      operator const T*() const { return &fArray->At(fIndex); }
-   };
+      Iterator_t operator+(std::ptrdiff_t n) const { return Iterator_t(fIndex + n, fArray); }
+
+      Iterator_t operator-(std::ptrdiff_t n) const
+      {
+         const auto index = std::ptrdiff_t(fIndex);
+         const auto newIndex = index >= n ? index - n : std::numeric_limits<decltype(fIndex)>::max();
+         return Iterator_t(newIndex, fArray);
+      }
+
+      std::ptrdiff_t operator-(const Iterator_t &other) const { return fIndex - other.fIndex; }
+
+      Iterator_t &operator+=(std::ptrdiff_t n)
+      {
+         auto &self = *this;
+         self = self + n;
+         return self;
+      }
+
+      Iterator_t &operator-=(std::ptrdiff_t n)
+      {
+         auto &self = *this;
+         self = self - n;
+         return self;
+      }
 
-   typedef Iterator_t iterator;
+      bool operator<(const Iterator_t &other) const { return fIndex < other.fIndex; }
+      bool operator>(const Iterator_t &other) const { return fIndex > other.fIndex; }
+      bool operator<=(const Iterator_t &other) const { return !(*this > other); }
+      bool operator>=(const Iterator_t &other) const { return !(*this < other); }
 
-   TTreeReaderArray(TTreeReader& tr, const char* branchname):
-      TTreeReaderArrayBase(&tr, branchname, TDictionary::GetDictionary(typeid(T)))
-   {
-      // Create an array reader of branch "branchname" for TTreeReader "tr".
-   }
+      T &operator[](std::size_t index) { return *(*this + index); }
+      const T &operator[](std::size_t index) const { return *(*this + index); }
+
+      operator T *() { return &fArray->At(fIndex); }
+      operator const T *() const { return &fArray->At(fIndex); }
+
+      bool IsValid() const { return fArray != nullptr; }
+   };
 
-   T& At(size_t idx) { return *(T*)UntypedAt(idx); }
-   T& operator[](size_t idx) { return At(idx); }
+   using iterator = Iterator_t;
+   using const_iterator = const Iterator_t;
```

Comment:
Doesn't that mean that this is ill-formed:
```
auto i = arr.cbegin();
i += 3;
```
The "const" part relates to modifying the elements, not the iterator.

Reply:
As per my comment below, I got `const_iterator` wrong. In the process of fixing it :)